Augment device object with dashboard url

lekkas opened the pull request:
Closes #225

pcarranzav commented on tests/integration.coffee:
```diff
@@ -48,6 +48,9 @@ if not _.every [
 	console.error('Missing environment credentials')
 	process.exit(1)
 
+buildDeviceDashboardUrl = (appId, deviceId) ->
```
Might make sense to include  and use this same function in device.coffee?

lekkas replied:
@pcarranzav yes, thank you for pointing out it makes perfect sense. I've updated the PR by adding a `resin.models.device.getDashboardUrl()` method cc @jviotti

emirotin commented on lib/models/device.coffee:
```diff
@@ -88,9 +90,33 @@ getDeviceModel = (deps, opts) ->
+
+			return "#{settings.get('dashboardUrl')}/apps/#{options.appId}/devices/#{options.deviceId}/summary"
+
 	addApplicationName = (device) ->
```
The method name is not appropriate anymore, please give it a better name, like `addExtraInfo` or smth like that

lekkas replied:
Done

lekkas commented:
@emirotin this is ready for re-review, the only issue now is that the PhantomJS tests are failing and I'm not sure why

emirotin commented:
Because `resin-settings-client` is node-only (it's using the filesystem) and shouldn't be used in the universal code.
The approach here I think is to add the optional dashboardUrl param, and when it's not set infer the dashboard URL from the API URL.

@pimterry your thoughts?

pimterry commented:
We could make `resin-settings-client` work in the browser too without too much trouble.

`Resin-settings-storage` already does both, switching LocalStorage client-side, and you could crib the implementation of that from there very easily (...which opens up some questions about why these are separate modules anyway, and whether more/everything could be integrated there...). However we do it, it'd be good for the future to be able to use this everywhere without worrying about it, and it lets us keep behaviour consistent between the web and node too, which will make everybody's lives easier.

pimterry commented on lib/models/device.coffee:
```diff
@@ -88,9 +88,32 @@ getDeviceModel = (deps, opts) ->
+	# @throws Exception if either appId or deviceId are empty
+	#
+	# @example
+	# dashboardUrl = resin.models.device.getDashboardUrl({ appId: '012345', deviceId: '678901' })
```
Can you use numbers instead of strings for the docs and all the examples here?

Doesn't matter if we happen to support both anyway, but we use numbers only for ids everywhere else in the SDK (we interpret strings as being UUIDs/app names, which we send to the API to get the corresponding ids when required). It would be good to be consistent with that, and having the documented API specifically require numbers leaves us open to potentially add UUID/app name support later or even https://github.com/resin-io/resin-sdk/issues/280 for total consistency, without it technically being a breaking change.

Maybe we should even mention that actually just to warn people: `@description Note that while passing strings for ids may work here, this will potentially break in future releases.`

lekkas replied:
@pimterry Updated the comments and docs to use require `Number`, also added an `isFinite` check for the passed appId/deviceId options.

+missing changelog entry.

emirotin commented:
We can make the settings client work in the browser, but it should do the (almost) completely different thing.

In Node it works like that: 
- the program (the CLI) ships with defaults (implemented by the settings-client)
- to change them we use the _environment_ - wither the env vars, or the persistent files on disk
- occasionally the same storage module used for that is also used for storing the token which does not belong to settings

In the browser it's different:
- we still want the persistence for the token
- we don't ship with defaults (but we actually can, by making the SDK-preconfigured universal which is easy to do)
- the environment is still there, but in the completely different form: it's the global props on `window` which is the conventional way to provide overrides to the 3rd-party libs

So what's the bottom line:
- we want the storage to be used as it is now for the token persistence in the universal manner
- we want the settings-client to be universal and treat the window-globals instead of the process.env in the browser (and do nothing to the files there, of course)
- IMO we shouldn't use the settings-client in the SDK itself (and let everything be passed as config opts to it), except the node-only settings persistence module
- but we want the SDK-preconfigured to become universal and read the params for the browser usage

pimterry commented:
That sounds good to me, you're right that window vars are more natural here. I do think we should namespace them though, to avoid conflicts. `apiUrl` will cause trouble otherwise! `window.resinSettings.X`?

I can also see the argument for pulling settings out if here entirely too, and doing that in the setup, e.g. SDK-preconfigured. I worry about it a little though because it means we'll have two places for defaults: once in resin-settings-client, with those values used in SDK-preconfigured, and once inside the SDK too, because otherwise we'll need to make all of these config arguments required for every SDK usage. Seems like that could get messy, but I like the approach overall, it's just how we do it. I think that can wait for now anyway, so we should fix settings-client for this first, merge this, and then later on move the usage elsewhere to avoid a direct settings dep here. Sound good?

emirotin commented:
Yes, I would call them window.RESIN_SETTINGS = { apiUrl: '...' } or smth
like that.

I think we need to decouple the defaults logic from the full settings
retrieval logic, and reuse it in both places.
And yes, the short-term plan sounds good.

On Tue, Apr 25, 2017 at 11:07 AM, Tim Perry <notifications@github.com>
wrote:

> That sounds good to me, you're right that window vars are more natural
> here. I do think we should namespace them though, to avoid conflicts.
> apiUrl will cause trouble otherwise! window.resinSettings.X?
>
> I can also see the argument for pulling settings out if here entirely too,
> and doing that in the setup, e.g. SDK-preconfigured. I worry about it a
> little though because it means we'll have two places for defaults: once in
> resin-settings-client, with those values used in SDK-preconfigured, and
> once inside the SDK too, because otherwise we'll need to make all of these
> config arguments required for every SDK usage. Seems like that could get
> messy, but I like the approach overall, it's just how we do it. I think
> that can wait for now anyway, so we should fix settings-client for this
> first, merge this, and then later on move the usage elsewhere to avoid a
> direct settings dep here. Sound good?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-sdk/pull/226#issuecomment-296953152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgGCM7f6-T4lWKKrn5eLxid8cXOlv-zks5rzanTgaJpZM4KHUe2>
> .
>



-- 
Eugene Mirotin
Senior Frontend Engineer
site: Resin.io <https://resin.io/>, twitter: @resin_io
<https://twitter.com/resin_io>

pimterry commented:
@lekkas Sounds like we have a plan. I'm assuming the above makes sense and you're happy to finish this up, but just shout if not, or if you need a hand at all.

I've filed a separate issue for dropping the resin-settings-client dep entirely over here: https://github.com/resin-io/resin-sdk/issues/310. We can do that independently later on.

lekkas commented:
@pimterry @emirotin I've updated the PR accordingly, please let me know if the approach was what you expected.

lekkas commented:
Hm tests are failing with an interesting ` ResinDeviceNotFound: Device not found: 3af33a32bf2962e4cd4254ac5bc9ee99683a3a926697918b9cc23075211712` error. Does anyone have access to the test account we are using?

lekkas commented on lib/models/device.coffee:
```diff
@@ -51,6 +51,10 @@ getDeviceModel = (deps, opts) ->
 
+	# Infer dashboardUrl from apiUrl if former is undefined
+	if not dashboardUrl?
+		dashboardUrl = apiUrl.replace(/api/, 'dashboard')
```
Should we just throw an exception here if `apiUrl` is not set?

emirotin replied:
The apiUrl is required param, we should throw (if we don't) as early as possible - in the very beginning of the SDK factory.

lekkas replied:
Ok so `apiUrl` is set as a default options if not passed to the SDK factory, so I guess we are fine here: https://github.com/resin-io/resin-sdk/blob/master/lib/resin.coffee#L60-L65

emirotin commented:
> Hm tests are failing with an interesting

Does it happen consistently? We've seen this sporadically and still didn't figure out what's going on.

lekkas commented:
> Does it happen consistently? We've seen this sporadically and still didn't figure out what's going on.

@emirotin I guess I just hit that then, the tests are passing now 👍

emirotin commented:
This needs to be rebased and the changelog entry should be outside of the recently added"v6.1.0" header

emirotin commented on lib/models/device.coffee:
```diff
@@ -51,6 +51,10 @@ getDeviceModel = (deps, opts) ->
 	exports = {}
 
+	# Infer dashboardUrl from apiUrl if former is undefined
+	if not dashboardUrl?
```
Hm, interesting. We do actually have this logic in resin-settings-client: https://github.com/resin-io-modules/resin-settings-client/blob/master/lib/defaults.coffee#L73
OK let's merge this as is and update it when #310 is implemented.

emirotin commented:
Also you seem to have only rebuild the node part, please run full `npm run build`

pimterry commented on lib/models/device.coffee:
```diff
@@ -88,9 +92,33 @@ getDeviceModel = (deps, opts) ->
+	###
+	exports.getDashboardUrl = getDashboardUrl = (options = {}) ->
+		isFinite = require('lodash/isFinite')
+		url = require('url')
```
Nit: we normally do requires at the top level in the SDK, it'd be good to keep everything together.

You're right that we do this in the CLI though, and I'm now wondering whether we should in the SDK for startup performance too... Not right now though :-).

pimterry replied:
I've filed an issue to think about this for the future: https://github.com/resin-io/resin-sdk/issues/314

lekkas replied:
@pimterry makes sense, I've bubbled the `require`'s up since this is how it's done currently

lekkas commented:
> Also you seem to have only rebuild the node part

Ah right, I thought `resin-browser` built JS is committed pre-publish, also rebased and fixed CHANGELOG @emirotin

emirotin commented:
It is, but we also publish it to GH so include it in the builds
